### PR TITLE
use localStorage to persist binder pods across pages

### LINKED ISSUE #224: Persist a server connection across pages
Many times, a user switches pages within the same domain and wants to run some more code. Right now, Thebelab will request a **new** binder session each time this happens, which puts burden on mybinder.org.

It would be great if we stored the Binder session information to re-use the session if it exists. Then a user would get super fast sessions in subsequent pageloads.

@ines kindly showed us where she does this in Juniper, I think we should follow the same pattern!

https://github.com/ines/juniper/blob/master/src/index.js#L57-L64

### PR DESCRIPTION
**Pull request by rkevin-arch**
This should resolve #224 (we can make a new issue for the persisting kernel (not binder pod) part if needed).

When Binder spawns a pod, the jupyter pod URL and token are saved in the browser's localStorage. When Thebe is activated and there is already a pod URL+token in localStorage, it will try to reuse the existing pod. If it fails, then it falls back to creating a new pod.

### CODE REVIEW

**Comment by minrk** [comment]
Great! I don't *think* persisting the kernel is as valuable as the server (kernel starts are fast, server starts with Binder are quite slow), so we can wait until someone requests that specifically.

**Comment by rkevin-arch** [comment]
I agree. I also think persisting kernels across pages can be potentially confusing, since code run on the first page might affect stuff in the second page. If someone thinks there's a valid use case, we can deal with it then.

File: src/thebelab.js [hunk 1]
```diff
@@ -436,6 +469,21 @@ export function requestBinder({
           break;
         case "ready":
           es.close();
+          try {
+            // save the current connection url+token to reuse later
+            window.localStorage.setItem("thebe_binder_pod_url_" + url, msg.url);
```
Comment (minrk) [inline src/thebelab.js:474]: maybe a single item with the two fields as JSON? e.g. info 


```suggestion
            window.localStorage.setItem("thebe-binder-" + url, JSON.stringify({url: msg.url, token: msg.token, started: new Date()}));
```

then `started` can also be used to ignore saved data older than e.g. one day

File: src/thebelab.js [hunk 2]
```diff
@@ -393,11 +393,44 @@ export function requestBinder({
...
+  return new Promise(async (resolve, reject) => {
+    // if binder already spawned our pod and we remember the creds, reuse it
+    let spawned_pod_url = window.localStorage.getItem(
+      "thebe_binder_pod_url_" + url
+    );
+    let spawned_pod_token = window.localStorage.getItem(
+      "thebe_binder_pod_token_" + url
+    );
+    if (
+      ignoreSavedPod === false &&
```
Comment (minrk) [inline src/thebelab.js:406]: ignoreSavedPod is undefined?

File: src/thebelab.js [hunk 3]
```diff
@@ -393,11 +393,44 @@ export function requestBinder({
...
   console.log("Binder build URL", url);
-  events.trigger("status", {
-    status: "building",
-    message: "Requesting build from binder",
-  });
-  return new Promise((resolve, reject) => {
+
+  return new Promise(async (resolve, reject) => {
+    // if binder already spawned our pod and we remember the creds, reuse it
+    let spawned_pod_url = window.localStorage.getItem(
```
Comment (minrk) [inline src/thebelab.js:399]: Let's try to avoid kubernetes names in here, and call it e.g. "existing_server_url"

**Review by minrk — COMMENTED** [review]
Awesome! This seems great. A few comments about details, but this looks very nice.

File: src/thebelab.js (hunk 2 above)
Comment (rkevin-arch) [reply src/thebelab.js:406]: Whoops, remnant of some other code that no longer exists. Removed.

**Comment by rkevin-arch** [comment]
I've made some changes according to your comments. I also made it so the saved info is not used after 86400 seconds (a day). Probably should change it to an hour or leave it configurable. Let me know what you think.

**Comment by choldgraf** [comment]
Really nice - thanks for this! I agree about not worrying about persisting kernels, since jupyter has a "1 kernel per notebook" mental model, I think most users will also assume "1 kernel per webpage"

**Comment by rowanc1** [comment]
This works well for me!

For anyone curious on seeing or clearing the local storage, you can do that in Chrome here:
image

File: src/thebelab.js [hunk 4]
```diff
@@ -393,11 +393,40 @@ export function requestBinder({
...
+      try {
+        await Kernel.listRunning(settings);
+        console.log("Saved binder pod is valid, reusing connection");
+        resolve(settings);
+        return;
+      } catch (err) {
+        console.log(
+          "Saved binder pod info seems to be invalid, respawning pod",
+          err
+        );
```
Comment (rowanc1) [inline src/thebelab.js:421]:
```suggestion
        console.log(
          "Saved binder pod info seems to be invalid, respawning pod",
          err
        );
        localStorage.removeItem("thebe-binder-" + url);
```

**Review by rowanc1 — APPROVED** [review]
Have looked over the PR, and tested locally - including trying to screw up URLs and keys. :)

**Comment by moorepants** [comment]
Sounds like this is good from the three reviewers. Gonna merge. Thanks @rkevin-arch!

**Comment by choldgraf** [comment]
This is very exciting!

**Comment by minrk** [comment]
This is rad! I added a little configuration to it in #280

**Pull request merged**